synse/http_integration_test.go

```diff
@@ -0,0 +1,551 @@
+	}
+
+	client, err := NewHTTPClientV3(&Options{
+		Address: "localhost:5000",
+	})
+	assert.NotNil(t, client)
+	assert.NoError(t, err)
+
+	opts := scheme.ScanOptions{}
+	devices, err := client.Scan(opts)
+	assert.NoError(t, err)
+	assert.Equal(t, 22, len(devices))
+
+	for _, device := range devices {
+		assert.NotEmpty(t, device)
+	}
+
+	for _, device := range devices {
+		assert.NotEmpty(t, device)
+
+		info, err := client.Info(device.ID)
+		assert.NoError(t, err)
+		assert.NotEmpty(t, info.Timestamp)
+		assert.NotEmpty(t, info.ID)
+		assert.NotEmpty(t, info.Type)
+		assert.NotEmpty(t, info.Plugin)
+		assert.NotEmpty(t, info.Capabilities.Mode)
+		assert.NotEmpty(t, info.Tags)
+		assert.NotEmpty(t, info.Metadata)
+		assert.Empty(t, info.Capabilities.Write.Actions)
+
+		// TODO - add sort_index to the scheme
+
+		// only led devices have alias
+		if info.Type == "led" {
+			assert.NotEmpty(t, info.Alias)
+		} else {
+			assert.Empty(t, info.Alias)
+		}
+
+		for _, output := range info.Outputs {
+			assert.NotEmpty(t, output.Name)
+			// assert.NotEmpty(t, output.Type) // FIXME - only airflow types are empty?
+
+			// led and lock devices don't produce unit and precision output
+			if output.Name == "state" || output.Name == "color" || output.Name == "status" {
+				assert.Empty(t, output.Unit.Name)
+				assert.Empty(t, output.Unit.Symbol)
+				assert.Empty(t, output.Precision)
+			} else {
+				assert.NotEmpty(t, output.Unit.Name)
+				assert.NotEmpty(t, output.Unit.Symbol)
+				assert.NotEmpty(t, output.Precision)
+			}
+
+			assert.Empty(t, output.ScalingFactor)
+		}
+	}
+}
+
+func TestIntegration_Read(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+
+	client, err := NewHTTPClientV3(&Options{
+		Address: "localhost:5000",
+	})
+	assert.NotNil(t, client)
+	assert.NoError(t, err)
+
+	opts := scheme.ReadOptions{}
+	devices, err := client.Read(opts)
+	assert.NoError(t, err)
+	assert.Equal(t, 25, len(devices))
+
+	for _, device := range devices {
+		assert.NotEmpty(t, device.Device)
+		assert.NotEmpty(t, device.Timestamp)
+		// assert.NotEmpty(t, device.Type) // FIXME - fan and airflow types are empty?
+		assert.NotEmpty(t, device.DeviceType)
+
+		// led and lock devices don't produce unit output
+		if device.DeviceType == "led" || device.DeviceType == "lock" {
+			assert.Empty(t, device.Unit.Name)
+			assert.Empty(t, device.Unit.Symbol)
+		} else {
+			assert.NotEmpty(t, device.Unit.Name)
+			assert.NotEmpty(t, device.Unit.Symbol)
+		}
+
+		// NOTE - device.Value could be 0 so no need to check that
+	}
+}
+
+func TestIntegration_ReadDevice(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+
+	client, err := NewHTTPClientV3(&Options{
+		Address: "localhost:5000",
+	})
+	assert.NotNil(t, client)
+	assert.NoError(t, err)
+
+	opts := scheme.ScanOptions{}
+	devices, err := client.Scan(opts)
+	assert.NoError(t, err)
+	assert.Equal(t, 22, len(devices))
+
+	for _, device := range devices {
```

Comment:
I'm thinking that it may be worthwhile to define a device configuration for the emulator for these tests so we know the set of devices we are working with that won't change. With that and the fact that we have deterministic IDs, we shouldn't have to iterate over scan results to test ReadDevice (and I think the Info test does the same thing) for every device. Instead we can pick one and verify that the values we get back match the expected. This allows us to use more specific checks than just `NotEmpty` which seems desirable